src/Chainweb/Pact/PactService.hs

```diff
@@ -827,8 +830,16 @@ execPreInsertCheckReq txs = pactLabel "execPreInsertCheckReq" $ withDiscardedBat
     withCurrentCheckpointer "execPreInsertCheckReq" $ \pdb -> do
       let v = _chainwebVersion psEnv
           cid = _chainId psEnv
-      liftIO $ fmap Discard $
-        validateChainwebTxs logger v cid cp parentTime currHeight txs (runGas pdb psState psEnv)
+          timeoutLimit = fromInteger $ toMicroseconds $ _psPreInsertCheckTimeout psEnv
+          act = fmap Discard $
+              validateChainwebTxs logger v cid cp parentTime currHeight txs (runGas pdb psState psEnv)
+
+      liftIO $ timeout timeoutLimit act >>= \case
+        Just r -> pure r
+        Nothing -> do
+            mpaBadlistTx mpAccess (V.map requestKeyToTransactionHash requestKeys)
+            throwM $ PreInsertCheckTimeout requestKeys timeoutLimit
+
```

Comment:
I don't think we actually have to throw an error here, just return `InsertError`.